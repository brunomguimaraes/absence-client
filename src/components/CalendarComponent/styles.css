.CalendarContainer {
  height: 560px;
}