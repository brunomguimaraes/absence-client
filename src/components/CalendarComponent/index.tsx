import React, { useEffect, useState } from 'react';
import { Calendar, momentLocalizer } from 'react-big-calendar';
import moment from 'moment'

import './styles.css';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import api from '../../services/api';

type AbscenceType = "sickness" | "vacation";

interface Member {
  id: number,
  name: string,
  userId: number,
}

interface Absence {
  id: number;
  userId: number;
  type: AbscenceType;
  startDate: Date;
  endDate: Date;
}

interface Event {
  userId: number;
  name: string;
  start: Date;
  end: Date;
  title: string;
}

const CalendarComponent = () => {
  const [members, setMembers] = useState<Member[]>([]);
  const [events, setEvents] = useState<Event[]>([]);
  const [absences, setAbsences] = useState<Absence[]>([]);
  const localizer = momentLocalizer(moment);

  useEffect(() => {
    api.get('members').then(response => {
      setMembers(response.data);
    })
    api.get('absences').then(response => {
      setAbsences(response.data);
    })
  }, []);

  useEffect(() => {
    if (members && absences) {
      const eventWithTitle = absences.map(absence => ({
        start: absence.startDate,
        end: absence.endDate,
        userId: absence.userId,
        title: `${members.find(
          member => member.userId === absence.userId)!.name} is 
          ${absence.type === "vacation" ?
          "on vacation"
          : "sick"}`,
        name: members.find(member => member.userId === absence.userId)!.name,
      }));

      setEvents(eventWithTitle);
    }
  }, [members, absences]);

  return (
    <div className={"CalendarContainer"}>
      {events && <Calendar
        localizer={localizer}
        events={events}
        defaultDate={new Date(2017, 0, 1)}
        startAccessor="start"
        endAccessor="end"
      />}
    </div>
  )
}

export default CalendarComponent;