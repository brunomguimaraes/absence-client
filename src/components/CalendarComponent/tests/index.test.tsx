import React from 'react';
import { shallow, mount } from 'enzyme';
import CalendarComponent from '..';
import { setHookState } from '../../../utils/test';

const reactMock = require('react')

describe('CalendarComponent', () => {
  let wrapper;

  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      }))
    });
  });

  it('renders the component', () => {
    wrapper = shallow(<CalendarComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the table with events', () => {
    wrapper = mount(<CalendarComponent />);

    reactMock.useState = setHookState({
      events: ["a lot of events"],
    })

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Calendar')).toMatchSnapshot();
    expect(wrapper.find('Calendar').exists()).toEqual(true)
  });
});
