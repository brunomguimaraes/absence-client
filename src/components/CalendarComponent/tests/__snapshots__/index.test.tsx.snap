// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarComponent renders the component 1`] = `ShallowWrapper {}`;

exports[`CalendarComponent renders the table with events 1`] = `ReactWrapper {}`;

exports[`CalendarComponent renders the table with events 2`] = `ReactWrapper {}`;
